src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -29,33 +24,15 @@ namespace Microsoft.CodeAnalysis.Completion
     /// <summary>
     /// A subtype of <see cref="CompletionService"/> that aggregates completions from one or more <see cref="CompletionProvider"/>s.
     /// </summary>
-    public abstract partial class CompletionServiceWithProviders : CompletionService, IEqualityComparer<ImmutableHashSet<string>>
+    public abstract partial class CompletionServiceWithProviders : CompletionService
     {
-        private readonly object _gate = new();
```

Comment:
i'm not quite certain why all this code had to move.  Why not just keep it here, but compute it on hte BG?

Reply:
I don't like mixing the logic of getting all the providers with implementation of CompletionService

Reply:
i suppose CompletoinServiceWithProviders was alwyas about the providers for me :)  so it made sense for all the provider logic to be in it.